aya/src/programs/links.rs

```diff
@@ -28,6 +28,17 @@ pub trait Link: std::fmt::Debug + 'static {
     fn detach(self) -> Result<(), ProgramError>;
 }
 
+bitflags::bitflags! {
```

Comment:
I see where you're going.  I took a look, and I believe the use of bitflags is the more usable interface in this case.  This allows the caller to easily set a combination of flags if they wish.  The bitflags API frees the developer from having to do bit manipulation themselves.

If we later choose to add support for more flags, such as `BPF_F_REPLACE`, `BPF_F_BEFORE`, and `BPF_F_AFTER`, then we would add the ability to supply the `bpf_attr` anonymous struct's three additional fields of `replace_bpf_fd`, `relative_fd`, `relative_id` – those would be separate arguments.  We would also add a way to set the `BPF_F_ID` flag to distinguish between the latter two.

I _think_ the use of bitflags now is inline with where we would want to go.

Reply:
> This allows the caller to easily set a combination of flags if they wish. The bitflags API frees the developer from having to do bit manipulation themselves.

My point is that I suspect we won't ever expose the bitflag to the aya API.

> If we later choose to add support for more flags, such as BPF_F_REPLACE, BPF_F_BEFORE, and BPF_F_AFTER, then we would add the ability to supply the bpf_attr anonymous struct's three additional fields of replace_bpf_fd, relative_fd, relative_id – those would be separate arguments.

Exactly, similar to https://docs.aya-rs.dev/aya/programs/xdp/struct.xdp#method.attach_to_link where we use a REPLACE bit, but don't expose it to the API, see https://docs.aya-rs.dev/src/aya/programs/xdp.rs#233

Or in other words: you have to pass (flag, fd_or_id) and it's unlikely that we'll have a single method to do that, but more realistically we'll have attach_replace(old) attach_relative(AttachFlags::Before, fd) etc, where the bitflags become an implementation detail but are not exposed

Reply:
These are good points.  From what I understand, you can also attach a program with both the `ALLOW_MULTI` and `ALLOW_OVERRIDE` flags set.  A bitflags argument is useful there to allow the caller to specify empty, either flag, or set both flags.  However, to keep the Aya API simple, we could go with your idea of an enum, and perhaps with a third value that represents both (`ALLOW_MULTI_AND_OVERRIDE`...?).  If we want to avoid an `EMPTY` value on the enum, we could include both an `attach` and `attach_with_flags` function.

An alternative is to implement the flags argument as a struct with two booleans:
```Rust
struct CgroupAttachFlags {
    allow_multi: bool,
    allow_override: bool,
}
```

A caller could then set either flag, both flags, or none.  That's probably my favorite option, as it seems clean.

Do any of these options resonate with you?

Reply:
from https://github.com/torvalds/linux/blob/35dfaad7188cdc043fde31709c796f5a692ba2bd/include/uapi/linux/bpf.h#L1094

```
 * BPF_F_ALLOW_OVERRIDE: If a sub-cgroup installs some bpf program,
 * the program in this cgroup yields to sub-cgroup program.
 *
 * BPF_F_ALLOW_MULTI: If a sub-cgroup installs some bpf program,
 * that cgroup program gets run in addition to the program in this cgroup.
 ```

so it seems like they're exclusive.

Given this and the docs of the other flags, I'm not convinced that at the aya level there's any need for bitflags, and bitflags are less idiomatic than regular enums so we should avoid them when possible.

Reply:
You're right, good instincts.  I've updated this change to use an enum, and included a function scoped to the crate to translate the enum to its BPF value.